### PR TITLE
Development 3.0: Bind stuff

### PR DESCRIPTION
**Pull request by cclerget**
**Description of the Pull Request (PR):**

Work in progress :
- Bind system
- Overlay/Underlay support
- Session management

### CODE REVIEW

File: src/pkg/util/fs/helper.go [hunk 1]
```diff
@@ -63,3 +64,28 @@ func IsSuid(name string) bool {
...
+func MkdirAll(path string, mode os.FileMode) error {
+	oldmask := syscall.Umask(0)
+	defer syscall.Umask(oldmask)
+
+	return os.MkdirAll(path, mode)
+}
+
+// RootDir returns the root directory of path (rootdir of /my/path is /my).
+// Returns "." if path is empty
+func RootDir(path string) string {
```
Comment (bauerm97) [inline src/pkg/util/fs/helper.go:79]: I think we should add a few sanity test cases for this function in `helper_test.go`

File: src/runtime/engines/common/oci/config.go [hunk 2]
```diff
@@ -30,6 +30,6 @@ func (c *Config) UnmarshalJSON(b []byte) error {
 	if err := json.Unmarshal(b, &c.Spec); err != nil {
 		return err
 	}
-	c.Generator = generate.NewFromSpec(&c.Spec)
+	c.Generator = generate.Generator{Config: &c.Spec}
```
Comment (bauerm97) [inline src/runtime/engines/common/oci/config.go:33]: Why is this change necessary? Just curious

**Review by bauerm97 — CHANGES_REQUESTED** [review]
Can we make sure that Makefile.stub is only installing the necessary dir? (Since I believe now everything happens under `$(SESSIONDIR)`)

Also this isn't directly from this PR, but `src/pkg/util/fs/layout/manager.go`, can you change `func (m *Manager) GetPath(path string) (string, error)` to use `filepath.Join()`? Thanks!

This is just a first pass review looking at some code. I have yet to fully review `container.go`. And I have yet to do testing on functionality.

File: src/pkg/util/fs/helper.go (hunk 1 above)
Comment (cclerget) [reply src/pkg/util/fs/helper.go:79]: Agreed

File: src/runtime/engines/common/oci/config.go (hunk 2 above)
Comment (cclerget) [reply src/runtime/engines/common/oci/config.go:33]: Eduardo has pointed that NewFromSpec is marked as deprecated https://github.com/opencontainers/runtime-tools/blob/699dbced9a52c2d60213d8f22d6e170840d69c06/generate/generate.go#L234

Comment (bauerm97) [reply src/runtime/engines/common/oci/config.go:33]: Good catch @ArangoGutierrez

**Comment by bauerm97** [comment]
Hey @cclerget I noticed a few issues so far in my testing:

- `--no-home` option leaves the `$HOME` environment variable set
- Config `allow pid ns = no` appears to not do anything
- In config file, at `LIMIT CONTAINER GROUPS` the default line says `DEFAULT: @LIMIT_CONTAINER_GROUPS_DEFAULT@` which should probably be `DEFAULT: NULL` same as limit container owners

**Review by bauerm97 — APPROVED** [review]
